### PR TITLE
[primTorch] refs: hsplit, vsplit

### PR DESCRIPTION
**Pull request by kshitij12345**
As per title

TODO:
* [x] Add error inputs (already exist)

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -1732,7 +1732,7 @@ def merge_dicts(*dicts):
 
 Args:
     input (Tensor): tensor to split.
-    indices_or_sections (Tensor, int or list or tuple of ints): See argument in :func:`torch.tensor_split`.
```
Comment (kshitij12345) [inline torch/_torch_docs.py:1735]: Does not accept tensor
https://github.com/pytorch/pytorch/blob/8ad305f375bf20e88398f50b48425a3a7a80b685/aten/src/ATen/native/native_functions.yaml#L4407-L4411

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -1822,6 +1824,73 @@ def tensor_split(
...
+            msg = (
+                "torch.hsplit attempted to split along dimension "
+                + str(dim)
+                + ", but the size of the dimension "
+                + str(a.shape[dim])
+                + " is not divisible by the split_size "
+                + str(split_size)
+                + "!"
+            )
+            raise RuntimeError()
```
Comment (mruberry) [inline torch/_refs/__init__.py:1850]: msg should go here -- @ezyang added a helper function we should start using, too, see: 

https://github.com/pytorch/pytorch/blob/41e2611e6a3f53f549ad27585a58ba7f7c71594a/torch/_prims/utils.py#L1192

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -1822,6 +1824,73 @@ def tensor_split(
...
+        msg = (
+            "torch.hsplit requires a tensor with at least 1 dimension, but got a tensor with "
+            + str(a.ndim)
+            + " dimensions!"
+        )
+        raise RuntimeError(msg)
+    dim = 0 if a.ndim == 1 else 1
+    if isinstance(indices_or_sections, int):
+        split_size = indices_or_sections
+        if not (split_size != 0 and a.shape[dim] % split_size == 0):
```
Comment (mruberry) [inline torch/_refs/__init__.py:1840]: double negation here is a little hard to read -- maybe propagate the `not`?

Comment (mruberry) [reply torch/_refs/__init__.py:1840]: is this error message redundant with error checking done by tensor_split?

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -1822,6 +1824,73 @@ def tensor_split(
...
+            )
+            raise RuntimeError()
+        return tensor_split(a, split_size, dim)
+    elif isinstance(indices_or_sections, (list, tuple)):
+        split_sizes = indices_or_sections
+        return tensor_split(a, split_sizes, dim)
+    else:
+        raise TypeError(
+            "hsplit(): received an invalid combination of arguments. "
+            "Expected indices_or_sections to be of type int, list of ints or tuple of ints"
```
Comment (mruberry) [inline torch/_refs/__init__.py:1858]: "... but got type ..."

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -1822,6 +1824,73 @@ def tensor_split(
...
+                + " is not divisible by the split_size "
+                + str(split_size)
+                + "!"
+            )
+            raise RuntimeError()
+        return tensor_split(a, split_size, dim)
+    elif isinstance(indices_or_sections, (list, tuple)):
+        split_sizes = indices_or_sections
+        return tensor_split(a, split_sizes, dim)
+    else:
```
Comment (mruberry) [inline torch/_refs/__init__.py:1855]: No need for the else clause, this can just fall through to the error

Even nicer, for style points, is you can do:

```
if (int):

if not isinstance(indices_or_sections, (list, tuple)):
  error

fallthrough to case where it's a list or tuple
```

that would minimize indentation
```

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -1822,6 +1824,73 @@ def tensor_split(
...
+        split_size = indices_or_sections
+        if not (split_size != 0 and a.shape[0] % split_size == 0):
+            msg = (
+                "torch.vsplit attempted to split along dimension 0, but the size of the dimension "
+                + str(a.shape[0])
+                + " is not divisible by the split_size "
+                + str(split_size)
+                + "!"
+            )
+            raise RuntimeError(msg)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1882]: msg is here so that's cool but same questions as for hsplit

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -1732,7 +1732,7 @@ def merge_dicts(*dicts):
...
+    indices_or_sections (int or list or tuple of ints): See argument in :func:`torch.tensor_split`.
```
Comment (mruberry) [inline torch/_torch_docs.py:1735]: Nice correction

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -1549,8 +1549,16 @@ def error_inputs_hsplit(op_info, device, **kwargs):
...
+
+    # Incorrect type for indices_or_section argument
+    err_msg3 = ("received an invalid combination of arguments.")
+    si3 = SampleInput(make_tensor((S, S, S),
+                                  dtype=torch.float32,
+                                  device=device),
+                      args=("abc",),)
     return (ErrorInput(si1, error_regex=err_msg1),
-            ErrorInput(si2, error_regex=err_msg2),)
+            ErrorInput(si2, error_regex=err_msg2),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1560]: I think we can just yield the ErrorInputs now, which might make this simpler and easier to read

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -1549,8 +1549,16 @@ def error_inputs_hsplit(op_info, device, **kwargs):
                                   dtype=torch.float32,
                                   device=device),
                       args=(0,),)
+
+    # Incorrect type for indices_or_section argument
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1553]: Nice additional case

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -19662,6 +19678,14 @@ def __init__(
             DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref'),
         )
     ),
+    PythonRefInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:19681]: Cool!

**Review by mruberry — COMMENTED** [review]
Overall looks good, @kshitij12345! A couple inline comments for your review

File: torch/_refs/__init__.py (hunk 3 above)
Comment (kshitij12345) [reply torch/_refs/__init__.py:1840]: I think this msg is better than the one that `tensor_split` throws

`tensor_split: number of sections must be greater than 0, but was 0` vs
`torch.hsplit attempted to split along dimension 1, but the size of the dimension 5 is not divisible by the split_size 0!`

File: torch/_refs/__init__.py [hunk 11]
```diff
@@ -1822,6 +1824,64 @@ def tensor_split(
...
+            + "!"
+        )
+        check((split_size != 0 and a.shape[dim] % split_size == 0), lambda: msg)
+        return tensor_split(a, split_size, dim)
+    if not isinstance(indices_or_sections, (list, tuple)):
+        raise TypeError(
+            "hsplit(): received an invalid combination of arguments. "
+            "Expected indices_or_sections to be of type int, list of ints or tuple of ints "
+            f"but got type {type(indices_or_sections)}"
+        )
```
Comment (kshitij12345) [inline torch/_refs/__init__.py:1855]: Would be nice if `check` can also take the type of error to raise. This would be changed to
`check(isinstance(indices_or_sections, (list, tuple)), msg, error_type=TypeError)`?

Comment (mruberry) [reply torch/_refs/__init__.py:2088]: cc @ezyang -- we're already seeing variants of `check` pop up for different error types

File: torch/_refs/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_refs/__init__.py]: OK

File: torch/_refs/__init__.py [hunk 12]
```diff
@@ -2055,6 +2057,64 @@ def tensor_split(
...
+    msg = (
+        "torch.hsplit requires a tensor with at least 1 dimension, but got a tensor with "
+        + str(a.ndim)
+        + " dimensions!"
+    )
+    check(a.ndim >= 1, lambda: msg)
+    dim = 0 if a.ndim == 1 else 1
+    if isinstance(indices_or_sections, int):
+        split_size = indices_or_sections
+        msg = (
```
Comment (mruberry) [inline torch/_refs/__init__.py:2072]: Doing final checks -- what I don't like about this pattern is that we're constructing the message even when we don't error, cc @ezyang -- I think we have to check for the condition to construct the message only in the error state, then if we want to use `check` just `check(False, ...`

File: torch/_refs/__init__.py [hunk 13]
```diff
@@ -2055,6 +2057,64 @@ def tensor_split(
         return tuple(splits)
 
 
+def hsplit(
+    a: TensorLikeType, indices_or_sections: DimsType
+) -> Tuple[TensorLikeType, ...]:
+    msg = (
```
Comment (mruberry) [inline torch/_refs/__init__.py:2063]: Same issue here about guarding message construction on the error

File: torch/_refs/__init__.py (hunk 12 above)
Comment (mruberry) [reply torch/_refs/__init__.py:2072]: (the other thing to do is we could define a function that returns the string, which would also delay string construction and which `check` does accept. I'm not sure how costly defining a function is, however -- I'd prefer we hide as much as possible behind the error checks

File: torch/_refs/__init__.py (hunk 11 above)
Comment (ezyang) [reply torch/_refs/__init__.py:1977]: Optional third argument sounds good to me

File: torch/_refs/__init__.py (hunk 12 above)
Comment (ezyang) [reply torch/_refs/__init__.py]: This is why check takes a lambda; `check(b, lambda: string_construction)`. Taking a string directly is wrong and won't actually work. I checked and dynamo is able to trace past the lambda construction.

File: torch/_refs/__init__.py [hunk 14]
```diff
@@ -2090,6 +2092,88 @@ def tensor_split(
         return tuple(splits)
 
 
+def hsplit(
```
Comment (mruberry) [inline torch/_refs/__init__.py:2095]: Nit: I feel like a common helper could have implemented these but nbd

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by mruberry** [comment]
This just needs a rebase to resolve the merge conflict, @kshitij12345

**Comment by kshitij12345** [comment]
Onnx failure looks unrelated to the PR. See https://github.com/pytorch/pytorch/issues/78844

@pytorchbot merge
